### ISSUE TITLE
Refresh / Reload song selection (shortcut)

### ISSUE THREAD

**Issue by natrius** [UltraStar-Deluxe/USDX#549]
### Desired behaviour

I'm right now building the database and when addings songs i want to check if everything is working properly. Sometimes i have to /want to change quite some stuff that results in a weird looking or just partly working song until i restart Ultrastar Deluxe.
Its not a big problem, but it would be a nice feature to be able to reload or refresh the song selection without having to restart all the time.

**Comment by piegamesde** [comment]
See also #494

**Comment by kiliankoe** [comment]
Just found this issue while looking for such a feature, it'd definitely be super nice to have. My friends and I usually add several songs when playing and have to restart Ultrastar every time 😅 But reading the other issue it does sound like it would be quite a bit of work and while the benefit would be nice, it's probably not worth the effort required 🙃

**Comment by jc0rn3t** [comment]
loading the songs in a playlist might help ? You can quite easily reload the playlist and have this updated in your songs view.

**Comment by kiliankoe** [comment]
I might've misunderstood the conversation previously, but I thought this to refer to songs added to the library that Ultrastar doesn't yet know about, so I wouldn't be able to add them to a playlist either, or how do you mean?

**Comment by jc0rn3t** [comment]
My bad, you're right. Trying to load that into a playlist wouldn't work :/

**Comment by barbeque-squared** [comment]
@kiliankoe is as far as I understand it correct.

What we're trying to address here is (for example) that you've copied in some songs while USDX was running. Right now you need to exit the game + restart it completely. Or maybe some of the txt's had errors like pointing to a non-existing audio file.

My one issue with reloading it like this is that it's a feature for the <1%. I too add batches of songs sometimes, but I just start it once, close it, then look in Error.log and fix the 90% in one go. As long as they show up in the song selection screen, USDX will reload _that individual song_ anyway if you play->Esc->Esc it, or go into the editor and `R`.
While I do think that the "help" offered in 494 isn't particularly useful (_why_ would I have to _manually_ re-assign allllll those extra things? why can't I just "this is the new songlist + call the `fixEverything()` function and boom, done"), I also don't really understand how one would "play" and "add songs" at the same time.

**Comment by kiliankoe** [comment]
> I also don't really understand how one would "play" and "add songs" at the same time.

In my specific setup the game is running on an external display (a projector), and I can edit the song library accordingly on the laptop that UltraStar is running on while others are singing a song.

> it's a feature for (I assume) <1% of the playerbase

very likely 😅

**Comment by dgruss** [comment]
@natrius you can do that with the usdx internal editor. just press E on any song. TAB shows a help screen to explain what all the keys do, but you can also use the mouse

independently, restarting the game may be less of a hassle when the loading time is lower, see #1121

**Comment by dgruss** [comment]
could be closed given that restarts are fast with #1121

**Comment by natrius** [comment]
I think "fast loadup time" is not the same as "continuous reload on change of files". But i also completely understand if thats not planned :)

I also like to point out - i have to edit almost every song a little bit. Thats why i'm asking this. Cannot imagine that close to nobody has to do nothing. 

Feel free to close the issue as "not planned". Thanks for all the great work you are doing! :)

**Comment by dgruss** [comment]
but editing works both within the game editor and when you edit a song file outside of the game and then select it in the game again and play it, it will use your updated notes and tags. so this already works!?

**Comment by natrius** [comment]
Oh, I'm sorry, i missed that, i will check, but if thats the case the issue can be closed obviously :D

**Comment by dgruss** [comment]
can be closed then

**Comment by SteveW94** [comment]
Hm...I have the exact same usecase as @kiliankoe 

And I am not sure, but coming with some songs you like and adding a lot of the ones your mates like during a session seemed like a pretty general usecase to me?
Casuse I thought nobody would already have all songs requested in his library and almost everytime dynamically adding of quite a bunch is requestet.

It might ve been a bit naive but I was hoping that hitting F5 in the sound-selection to reconstruct the soundselection once from the groundup but with a re-read of your songs folder wouldnt include so much code changes 🙈

**Comment by dgruss** [comment]
what's the advantage over restarting the game?

this would be clean to implement:
- F5 --> transition to loading screen --> transition back to song selection, state is lost 

but you could already have that behavior with:
- Alt+F4 --> Enter (to confirm) --> Enter (to start the game again or write a script which runs it in a loop) --> Enter twice to be back in song selection

adding this in code adds the risk of introducing new bugs when the sequence is already possible.

**Comment by SteveW94** [comment]
Well, It would be a pure convinience thing, nothing more nothing less...
In my case that means:

- Not navigating back and forth from notebook and projector screen with the mouse (using the mouse instead of the controller away from the laptop)
- I would expect to have the same song selected which I ve had selected before after the reload. To continue progressing my lib
- Not having again clicking  through game and profile selection every 10min when a new song comes in.

Something like that sums it up I guess

**Comment by dgruss** [comment]
I guess the better approach is to pre download all songs. It's a bit of disk space but the game can handle it.

**Comment by SteveW94** [comment]
We never know any songs before, like thats the whole point.

Every round is with a bunch of new people with different music taste and all.
Then they ask "Oh can I sing this ... or that ...?" Like you would do in a pub or sth.

The download aquivalent would be to download and keep in sync the whole https://usdb.animux.de/ DB with audios and videos, and I guess thats as well a bit too much to ask for getting a bit more flow and convinience 😆

**Comment by dgruss** [comment]
That is correct and that is what a lot of people do with usdb syncer. Just have the entire usdb locally.

**Comment by SteveW94** [comment]
I considered that more as a silly joke or sth tbh^^ 
I mean....with the current 80k sth songs, that would be...I dont know? Days of download time and terabytes of data to store locally.
The whole benefit of usdb-syncer and USDX in my current setup was that I can very lightwight carry around a initial setup with below 1GB on a stick or sth. and plug it into every laptop to start spontaniously and download new songs on demand in seconds.

But I can understand that this might be not the default use case for the majority here, and I dont wanna enforce any feature here, especially since the startup time got improved as mentioned.
It just sounded like a "Just call the reload folder method once again and refresh the screen" kinda thing to me, that was why it sounded like sth rather intuitive to do.
But when it is a hassle code wise I am fine with restarting the app as well, I was just courious what the thoughts behind this were. 😃

**Comment by bohning** [comment]
It's ~28k songs and ~1 TB of data with highest quality settings. The entire USDB fits on a tiny SDXC card.

**Comment by SteveW94** [comment]
Oh yeah, sorry about the number switch 28k songs.
But damn I actually expected far more space consumed by that....I have just looked and saw that a song with video consumes approx. 25MB....I expected it to have like...200-300MB with the video at first.... crazy 😮

Okay, if thats really possible to do that in one go (and somehow keep it synced when new songs are there as well), ist there a "short" way to trigger like a "Download All" in usdb-syncer?

And sorry for asking such kinda "naiv" questions here...I am coming back to ultrastar after years of pause where there was no usdb-syncer and the average microSD and USB-Stick size was idk...16-32GB maybe 64GB/128GB if you had the money so I never even thought of sth like download it all 😆

**Comment by bohning** [comment]
I wouldn't recommend downloading it all (ratio of unwanted to wanted songs gets quite bad), but no, there is no shortcut.
